clientgenv2/source_generator.go

```diff
@@ -76,6 +70,114 @@ func (rs ResponseFieldList) IsStructType() bool {
+		return rs[i].Name < rs[j].Name
+	})
+	return rs
+}
+
+type StructGenerator struct {
+	currentResponseFieldList ResponseFieldList
+	preMergedStructSources   []*StructSource
+	postMergedStructSources  []*StructSource
+}
+
+func NewStructGenerator(responseFieldList ResponseFieldList) *StructGenerator {
+	currentFields := make(ResponseFieldList, 0)
+	fragmentChildrenFields := make(ResponseFieldList, 0)
+	for _, field := range responseFieldList {
+		if field.IsFragmentSpread {
+			fragmentChildrenFields = append(fragmentChildrenFields, field.ResponseFields...)
+		} else {
+			currentFields = append(currentFields, field)
+		}
+	}
+
+	preMergedStructSources := make([]*StructSource, 0)
+	postMergedStructSources := make([]*StructSource, 0)
+
+	for _, field := range responseFieldList {
+		if field.IsFragmentSpread {
+			preMergedStructSources = append(preMergedStructSources, &StructSource{
+				Name: field.FieldTypeString(),
+				Type: field.ResponseFields.StructType(),
+			})
+		}
+	}
+
+	currentFields = mergeFieldsRecursively(currentFields, fragmentChildrenFields, &preMergedStructSources, &postMergedStructSources)
+	return &StructGenerator{
+		currentResponseFieldList: currentFields,
+		preMergedStructSources:   preMergedStructSources,
+		postMergedStructSources:  postMergedStructSources,
+	}
+}
+
+func mergeFieldsRecursively(targetFields ResponseFieldList, sourceFields ResponseFieldList, preMerged, postMerged *[]*StructSource) ResponseFieldList {
+	res := make(ResponseFieldList, 0)
+	targetFieldsMap := targetFields.MapByName()
+	for _, sourceField := range sourceFields {
+		if targetField, ok := targetFieldsMap[sourceField.Name]; ok {
+			if targetField.ResponseFields.IsBasicType() {
+				continue
+			}
+			*preMerged = append(*preMerged, &StructSource{
+				Name: sourceField.FieldTypeString(),
+				Type: sourceField.ResponseFields.StructType(),
+			})
+			*preMerged = append(*preMerged, &StructSource{
+				Name: targetField.FieldTypeString(),
+				Type: targetField.ResponseFields.StructType(),
+			})
+
+			targetField.ResponseFields = mergeFieldsRecursively(targetField.ResponseFields, sourceField.ResponseFields, preMerged, postMerged)
+			*postMerged = append(*postMerged, &StructSource{
+				Name: targetField.FieldTypeString(),
+				Type: targetField.ResponseFields.StructType(),
+			})
+		} else {
+			res = append(res, sourceField)
+		}
+	}
+	for _, field := range targetFieldsMap {
+		res = append(res, field)
+	}
+	res = res.SortByName()
+	return res
+}
+
+func (g *StructGenerator) MergedStructSources(sources []*StructSource) []*StructSource {
+	res := sources
+	// remove pre-merged struct
+	for _, preMergedTypeName := range g.preMergedStructSources {
+		for i, source := range res {
+			// when name is same, remove it
+			if source.Name == preMergedTypeName.Name {
+				res = append(res[:i], res[i+1:]...)
+				break
+			}
+		}
+	}
+
+	// append post-merged struct
+	res = append(res, g.postMergedStructSources...)
+
+	return res
+}
```

Comment:
In my opinion. I prefer this approach of creating a map first because it results in shallower nesting during lookups, making the code cleaner and easier to follow.
```suggestion
func structSourcesMapByTypeName(sources []*StructSource) map[string]*StructSource {
	structSourceMapByTypeName := make(map[string]*StructSource)
	for _, source := range sources {
		structSourceMapTypeName[source.Name] = source
	}

	return structSourceMapByTypeName
}
func (g *StructGenerator) MergedStructSources(sources []*StructSource) []*StructSource {
	preMargedStructSourceMapByTypeName := structSourcesMapByTypeName(g.preMergedStructSources)
	res := sources
	// remove pre-merged struct
	for i, source := range res {
		// when name is same, remove it
		if _, ok := preMargedStructSourceMapByTypeName[source.Name]; ok {
			res = append(res[:i], res[i+1:]...)
			break
		}
	}

	// append post-merged struct
	res = append(res, g.postMergedStructSources...)

	return res
}
```

Reply:
It's true. I use map to avoid double loop.